### ISSUE TITLE
CpSolverSolutionCallback.BooleanValue takes unreasonable long time with enough variables

### ISSUE THREAD

**Issue by johnchen902** [google/or-tools#1148]
Consider this Python 3 code (with ortools 6.10.6025):

```python
from ortools.sat.python import cp_model
import cProfile

class Callback(cp_model.CpSolverSolutionCallback):
    def __init__(self, variables):
        super().__init__()
        self.__vars = variables
        self.__count = 0

    def on_solution_callback(self):
        self.__count += 1
        for var in self.__vars:
            self.BooleanValue(var)
        if self.__count >= 10:
            self.StopSearch()

def main():
    model = cp_model.CpModel()
    var = [model.NewBoolVar(f'x_{i}') for i in range(10000)]
    model.Add(sum(var) == len(var) // 2)
    solver = cp_model.CpSolver()
    status = solver.SearchForAllSolutions(model, Callback(var))

if __name__ == '__main__':
    cProfile.run('main()')
```

Here is the profiler output (manually sorted and truncated):

```sh
         1410177 function calls in 34.016 seconds

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   34.016   34.016 {built-in method builtins.exec}
        1    0.004    0.004   34.016   34.016 <string>:1(<module>)
        1    0.001    0.001   34.013   34.013 quad.py:18(main)
        1    0.000    0.000   33.876   33.876 cp_model.py:1399(SearchForAllSolutions)
        1    0.302    0.302   33.876   33.876 {built-in method ortools.sat._pywrapsat.SatHelper_SolveWithParametersAndSolutionCallback}
       10    0.000    0.000   33.573    3.357 cp_model.py:1485(OnSolutionCallback)
       10    0.046    0.005   33.573    3.357 quad.py:11(on_solution_callback)
   100000    0.256    0.000   33.527    0.000 cp_model.py:1489(BooleanValue)
   100000    0.029    0.000   33.033    0.000 pywrapsat.py:217(Response)
   100000   33.004    0.000   33.004    0.000 {built-in method ortools.sat._pywrapsat.SolutionCallback_Response}
```

`Response` is only called in `BooleanValue` as a safety check:
```python
        if not self.Response().solution:
            raise RuntimeError('Solve() has not be called.')
```

While using `SolutionBooleanValue` is a quick workaround, this is clearly unacceptable.

**Comment by lperron** [comment]
Even if the tone is inappropriate, this is a real performance bug where I
was copying the solution over and over.
This is fixed on master and will be available in 7.0 (hopefully tomorrow
:-) ).

Thanks
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by johnchen902** [comment]
Sorry. I'm not sure how my tone is offending.

Thanks anyway.
